primo/utils/clustering_utils.py

```diff
@@ -145,3 +177,106 @@ def perform_clustering(wd: WellData, distance_threshold: float = 10.0):
     wd.col_names.register_new_columns({"cluster": "Clusters"})
 
     return clustered_data.n_clusters_
+
+
+def perform_louvain_clustering(
+    wd: WellData, distance_threshold: float = 10.0, cluster_threshold: int = 300
+) -> dict:
+    """
+    Partitions the data into smaller clusters using the Louvain community detection method,
+    limiting each well to connect only with its 10 closest neighbors within a threshold distance.
+    Dynamically adjusts the resolution parameter to ensure no cluster exceeds the size threshold.
+
+    Parameters
+    ----------
+    wd : WellData
+        Object containing well data
+
+    distance_threshold : float
+        Threshold distance (in miles) for breaking clusters
+
+    cluster_threshold : int
+        cluster size of the largest cluster
+
+    Returns
+    -------
+    dict
+        Dictionary of lists of wells contained in each cluster
+    """
+    # pylint: disable=too-many-locals
+    existing_clusters = _check_existing_cluster(wd)
+    if existing_clusters is not None:
+        return existing_clusters
```

Comment:
This returns an integer. Whereas if no existing clusters are found, you are returning a dict. agglomerative clustering is returning an integer. Please make consistent overall :-)

Reply:
Made both functions consistent. We now return a dictionary in both cases. We can access the number of clusters by accessing the keys of the dictionary.